File: Packages/Login/web/src/utils/fileRef.ts
export type StoredFileScheme = 'local' | 'oss' | 'external'

export interface StoredFileRef {
    scheme: StoredFileScheme
    path: string
}

export function parseStoredRef(raw: string | undefined | null): StoredFileRef | null {
    const value = raw?.trim()
    if (!value) return null

    if (value.startsWith('http://') || value.startsWith('https://')) {
        return { scheme: 'external', path: value }
    }
    if (value.startsWith('local:')) {
        return { scheme: 'local', path: value.slice(6) }
    }
    if (value.startsWith('oss:')) {
        return { scheme: 'oss', path: value.slice(4) }
    }

    return { scheme: 'local', path: value }
}

export function formatStoredRef(scheme: StoredFileScheme, path: string): string {
    const trimmedPath = path.trim()
    if (scheme === 'external') {
        return trimmedPath
    }
    return `${scheme}:${trimmedPath}`
}

function fileCdnBase(): string {
    const base = import.meta.env.VITE_FILE_CDN_BASE || import.meta.env.VITE_AVATAR_CDN_BASE || ''
    return base.replace(/\/$/, '')
}

export function resolveFileUrl(stored: string | undefined, apiBase: string): string | undefined {
    const ref = parseStoredRef(stored)
    if (!ref) return undefined

    if (ref.scheme === 'external') {
        return ref.path
    }

    if (ref.scheme === 'oss') {
        const cdnBase = fileCdnBase()
        return cdnBase ? `${cdnBase}/${ref.path}` : stored
    }

    const base = apiBase.replace(/\/$/, '')
    return `${base}/api/v1/user/avatar/${encodeURIComponent(ref.path)}`
}

File: Packages/Login/unit-auth/services/file_ref.go
package services

import (
	"errors"
	"strings"
)

const (
	SchemeLocal    = "local"
	SchemeOSS      = "oss"
	SchemeExternal = "external"
)

// StoredFileRef 存储引用（scheme + path，DB 层编码为 local:/oss:/https://）
type StoredFileRef struct {
	Scheme string
	Path   string
}

var ErrEmptyStoredRef = errors.New("empty stored file reference")

// ParseStoredRef 解析 DB 中的存储字符串
func ParseStoredRef(raw string) (StoredFileRef, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return StoredFileRef{}, ErrEmptyStoredRef
	}

	if strings.HasPrefix(raw, "http://") || strings.HasPrefix(raw, "https://") {
		return StoredFileRef{Scheme: SchemeExternal, Path: raw}, nil
	}
	if strings.HasPrefix(raw, "local:") {
		return StoredFileRef{Scheme: SchemeLocal, Path: strings.TrimPrefix(raw, "local:")}, nil
	}
	if strings.HasPrefix(raw, "oss:") {
		return StoredFileRef{Scheme: SchemeOSS, Path: strings.TrimPrefix(raw, "oss:")}, nil
	}

	// 历史裸文件名，视为 local
	return StoredFileRef{Scheme: SchemeLocal, Path: raw}, nil
}

// FormatStoredRef 格式化为 DB 存储字符串
func FormatStoredRef(scheme, path string) string {
	scheme = strings.TrimSpace(scheme)
	path = strings.TrimSpace(path)
	if scheme == SchemeExternal {
		return path
	}
	return scheme + ":" + path
}
